client/relay-chain-rpc-interface/src/reconnecting_ws_client.rs

```diff
@@ -0,0 +1,495 @@
+		))?;
+		Ok(rx)
+	}
+
+	/// Get a stream of all imported relay chain headers
+	pub fn get_imported_heads_stream(&self) -> Result<Receiver<RelayHeader>, RelayChainError> {
+		let (tx, rx) =
+			futures::channel::mpsc::channel::<RelayHeader>(NOTIFICATION_CHANNEL_SIZE_LIMIT);
+		self.send_register_message_to_worker(RpcDispatcherMessage::RegisterImportListener(tx))?;
+		Ok(rx)
+	}
+
+	fn send_register_message_to_worker(
+		&self,
+		message: RpcDispatcherMessage,
+	) -> Result<(), RelayChainError> {
+		self.to_worker_channel
+			.try_send(message)
+			.map_err(|e| RelayChainError::WorkerCommunicationError(e.to_string()))
+	}
+}
+
+/// Worker that should be used in combination with [`RelayChainRpcClient`]. Must be polled to distribute header notifications to listeners.
+struct ReconnectingWebsocketWorker {
+	ws_urls: Vec<Url>,
+	/// Communication channel with the RPC client
+	client_receiver: TokioReceiver<RpcDispatcherMessage>,
+
+	/// Communication channel with ourself
+	self_sender: TokioSender<RpcDispatcherMessage>,
+
+	/// Senders to distribute incoming header notifications to
+	imported_header_listeners: Vec<Sender<RelayHeader>>,
+	finalized_header_listeners: Vec<Sender<RelayHeader>>,
+	best_header_listeners: Vec<Sender<RelayHeader>>,
+}
+
+fn distribute_header(header: RelayHeader, senders: &mut Vec<Sender<RelayHeader>>) {
+	senders.retain_mut(|e| {
+				match e.try_send(header.clone()) {
+					// Receiver has been dropped, remove Sender from list.
+					Err(error) if error.is_disconnected() => false,
+					// Channel is full. This should not happen.
+					// TODO: Improve error handling here
+					// https://github.com/paritytech/cumulus/issues/1482
+					Err(error) => {
+						tracing::error!(target: LOG_TARGET, ?error, "Event distribution channel has reached its limit. This can lead to missed notifications.");
+						true
+					},
+					_ => true,
+				}
+			});
+}
+
+/// Manages the active websocket client.
+/// Responsible for creating request futures, subscription streams
+/// and reconnections.
+#[derive(Debug)]
+struct ClientManager {
+	urls: Vec<Url>,
+	active_client: Arc<JsonRpcClient>,
+	active_index: usize,
+}
+
+struct RelayChainSubscriptions {
+	import_subscription: Subscription<RelayHeader>,
+	finalized_subscription: Subscription<RelayHeader>,
+	best_subscription: Subscription<RelayHeader>,
+}
+
+/// Try to find a new RPC server to connect to.
+async fn connect_next_available_rpc_server(
+	urls: &Vec<Url>,
+	starting_position: usize,
+) -> Result<(usize, Arc<JsonRpcClient>), ()> {
+	tracing::debug!(target: LOG_TARGET, starting_position, "Connecting to RPC server.");
+	for (counter, url) in urls.iter().cycle().skip(starting_position).take(urls.len()).enumerate() {
+		let index = (starting_position + counter) % urls.len();
+		tracing::info!(target: LOG_TARGET, index, ?url, "Connecting to RPC node",);
+		if let Ok(ws_client) = WsClientBuilder::default().build(url).await {
+			return Ok((index, Arc::new(ws_client)))
+		};
+	}
```

Comment:
```suggestion
	for i in 0..urls.len() {
		let index = (starting_position + i) % urls.len();
		let url = &urls[index];
		tracing::debug!(target: LOG_TARGET, index, ?url, "Connecting to RPC node",);
		if let Ok(ws_client) = WsClientBuilder::default().build(url).await {
			return Ok((index, Arc::new(ws_client)))
		};
	}
```

Should be much faster (I'm not trusting the `skip` when `start_position` is big). I would also not print the connect message as info log.

Reply:
I think this one should stay `info!`. This is not really spammy and should only occur once the connection has been lost. In that case, I would like to see where the collator connects.